odl/trafos/wavelet.py

```diff
@@ -21,9 +21,836 @@
 from __future__ import print_function, division, absolute_import
 from future import standard_library
 standard_library.install_aliases()
+from builtins import range, str, super
 
 # External
+import numpy as np
 
 # Internal
+from odl.discr.lp_discr import DiscreteLp
+from odl.operator.operator import Operator
 
-__all__ = ()
+try:
+    import pywt
+    WAVELET_AVAILABLE = True
+except ImportError:
+    WAVELET_AVAILABLE = False
+
+__all__ = ('DiscreteWaveletTrafo', 'DiscreteWaveletTrafoAdjoint',
+           'DiscreteWaveletTrafoInverse', 'WAVELET_AVAILABLE')
+
+
+_SUPPORTED_IMPL = ('pywt',)
+
+
+def list_of_coeff_sizes(shape, nscale, wbasis, mode):
+    """Construct a size list from given wavelet coefficients.
+
+    Construct a list containing the sizes of the wavelet approximation
+    and detail coefficients when wavelet basis, number of scaling levels
+    and the shape of the original image are given.
+    Related to 2D and 3D multidimensional wavelet transform.
+
+    Parameters
+    ----------
+    shape : `tuple`
+        Number of pixels/voxels in the image. Its lenght must be 2 or 3.
+    nscale : `int`
+        Number of scales in the multidimensional wavelet
+        transform
+    wbasis : ``_pywt.Wavelet``
+        Describes properties of a selected wavelet basis
+    mode : `str`
+        Signal extention mode. Possible extension modes are
+        'zpd' - zero-padding: signal is extended by adding zero samples
+        'cpd' - constant padding: border values are replicated
+        'sym' - symmetric padding: signal extension by mirroring samples
+        'ppd' - periodic padding: signal is trated as a periodic one
+        'sp1' - smooth padding: signal is extended according to the
+        first derivatives calculated on the edges (straight line)
+        'per' - periodization: like periodic-padding but gives the
+        smallest possible number of decomposition coefficients.
+
+    Returns
+    -------
+    size_list : `list`
+        A list containing the sizes of the wavelet (approximation
+        and detail) coefficients at different scaling levels
+        size_list[0] = size of approximation coefficients at the coarsest level
+        size_list[1] = size of the detailed coefficients at the coarsest level
+        ...
+        size_list[n] = size of the detailed coefficients at the finest level
+        size_list[n+1] = size of original image
+        n = number of scaling levels = nscale
+
+    See also
+    --------
+    pywt : http://www.pybytes.com/pywavelets/
+    """
+    if len(shape) not in (2, 3):
+        raise ValueError('shape must have length 2 or 3, got {}.'
+                         ''.format(len(shape)))
+    P = np.zeros(shape)
+    max_level = pywt.dwt_max_level(shape[0], filter_len=wbasis.dec_len)
+    if nscale >= max_level:
+        raise ValueError('Too many scaling levels')
```

Comment:
See above - argument constraints should be documented.

The error message could be more informative, too. Especially since it may be hard to know how many levels one can use, it would be nice if the error message would state the maximum number, and along with it the provided number (in case when it is computed, not literally provided).

General advice: print the provided as well as the expected parameter (range). A good example is the shape message above.